src/shared/Core/CollectorWriter.cs

```diff
@@ -0,0 +1,172 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+using KnownGitCfg = GitCredentialManager.Constants.GitConfiguration;
+
+namespace GitCredentialManager
+{
+    public enum CollectorWriterState
+    {
+        Unknown = 0,
+        Stopped = 1,
+        Failing = 2,
+        Healthy = 3,
+    }
+
+    /// <summary>
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class CollectorWriter
+    {
+        private const int DefaultMaxQueueSize = 256;
+
+        private readonly Func<NamedPipeClientStream> _createPipeFunc;
+        private readonly BlockingCollection<string> _queue;
+
+        private Thread _writerThread;
+        private NamedPipeClientStream _pipeClient;
+        private bool _isDisposed;
+
+        public CollectorWriterState State { get; private set; } = CollectorWriterState.Unknown;
+
+        public CollectorWriter(Func<NamedPipeClientStream> createPipeFunc,
+            int maxQueueSize = DefaultMaxQueueSize)
+        {
+            EnsureArgument.NotNull(createPipeFunc, nameof(createPipeFunc));
+            EnsureArgument.Positive(maxQueueSize, nameof(maxQueueSize));
+
+            _createPipeFunc = createPipeFunc;
+            _queue = new BlockingCollection<string>(new ConcurrentQueue<string>(), boundedCapacity: maxQueueSize);
+        }
+
+        public void Start()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(CollectorWriter));
+            }
+
+            try
+            {
+                _writerThread = new Thread(BackgroundWriterThreadProc)
+                {
+                    Name = nameof(CollectorWriter),
+                    IsBackground = true
+                };
+
+                _writerThread.Start();
+                // Create a new pipe stream instance using the provided factory
+                NamedPipeClientStream pipe = _createPipeFunc();
+
+                // Specify an instantaneous timeout because we don't want to hold up the
+                // background thread loop if the pipe is not available.
+                pipe.Connect(timeout: 0);
+
+                _pipeClient = pipe;
```

Comment:
This commit makes me need to page-in all of the rules around `IDisposable` and this late copy-to-member is one of those things. Makes sense that we don't want to call `_pipeClient.Dispose()` if it failed to `Connect()` in the line above.

Reply:
I never have understood the rules of IDisposable in C#, so I'll defer to your freshly memory mapped pages.

Reply:
Thinking about it a day later, I'm not sure if I was right. The client's `Dispose()` should be resilient to a failed setup, so this precaution isn't necessary. In fact, the current code will not `Dispose()` the client if it fails to connect, which is not the correct thing to do. Perhaps we should be acting on `_pipeClient` and drop the local `pipe` variable?

Reply:
I'll admit to being a bit confused here. Is the consensus that we should be calling `Dispose` if the `Connect` call fails?

Reply:
I think the ideas is that you shouldn't have a method-local `pipe` variable and instead act on `_pipeClient` the entire time.